File: src/main/java/javabot/operations/InfoOperation.java
package javabot.operations;

import java.text.SimpleDateFormat;

import javabot.BotEvent;
import javabot.Javabot;
import javabot.Message;
import javabot.model.Factoid;
import javabot.dao.FactoidDao;
import org.springframework.beans.factory.annotation.Autowired;

/**
 * Simple operation to pull who added the factoid and when it was added
 */
public class InfoOperation extends BotOperation {
    @Autowired
    private FactoidDao dao;

    public InfoOperation(final Javabot bot) {
        super(bot);
    }

    @Override
    public boolean handleMessage(final BotEvent event) {
        final String message = event.getMessage().toLowerCase();
        final String channel = event.getChannel();
        boolean handled = false;
        if (message.startsWith("info ")) {
            final String key = message.substring("info ".length());
            final Factoid factoid = dao.getFactoid(key);
            if (factoid != null) {
                getBot().postMessage(new Message(channel, event,
                    String.format("%s was added by: %s on %s and has a literal value of: %s", key,
                        factoid.getUserName(), formatDate(factoid), factoid.getValue())));
            }
            getBot().postMessage(new Message(channel, event, "I have no factoid called \"" + key + "\""));
            handled = true;
        }
        return handled;
    }

    private String formatDate(final Factoid factoid) {
        return new SimpleDateFormat("MM-dd-yyyy' at 'K:mm a, z").format(factoid.getUpdated());
    }
}

File: src/main/kotlin/javabot/operations/InfoOperation.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import javabot.dao.FactoidDao
import java.time.format.DateTimeFormatter
import java.util.Locale
import javax.inject.Inject

/**
 * Simple operation to pull who added the factoid and when it was added
 */
class InfoOperation @Inject constructor(bot: Javabot, adminDao: AdminDao, var factoidDao: FactoidDao) :
        BotOperation(bot, adminDao)  {
    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        val message = event.value.lowercase(Locale.getDefault())
        if (message.startsWith("info ")) {
            val key = message.substring("info ".length)
            val factoid = factoidDao.getFactoid(key)
            if (factoid != null) {
                val formatter = DateTimeFormatter.ofPattern(INFO_DATE_FORMAT)
                val updated = factoid.updated
                val formatted = formatter.format(updated)
                responses.add(
                    Message(
                        event, Sofia.factoidInfo(
                            key, if (factoid.locked) "*" else "", factoid.userName,
                            formatted, factoid.value
                        )
                    )
                )
            } else {
                responses.add(Message(event, Sofia.factoidUnknown(key)))
            }
        }
        return responses
    }

    companion object {
        val INFO_DATE_FORMAT: String = "dd MMM yyyy' at 'KK:mm"
    }

}